### ISSUE TITLE
Add support for jump list

### ISSUE THREAD

**Issue by pengux** [junegunn/fzf.vim#865]
- Category
    - [ ] Question
    - [ ] Bug
    - [x] Suggestion
- OS
    - [x] Linux
    - [x] macOS
    - [x] Windows
    - [x] Etc.
- Vim
    - [x] Vim
    - [x] Neovim

I would like to request a feature to support listing the vim jump list using FZF. Just like for the list of buffers for example, you should be able to fuzzy search in the jump list and select and open that location.

**Comment by patricknraanes** [comment]
Here's my solution, inspired by
https://www.reddit.com/r/vim/comments/a3e5q7

    function! GetJumps()
      redir => cout
      silent jumps
      redir END
      return reverse(split(cout, "\n")[1:])
    endfunction
    function! GoToJump(jump)
        let jumpnumber = split(a:jump, '\s\+')[0]
        execute "normal " . jumpnumber . "\<c-o>"
    endfunction
    command! Jumps call fzf#run(fzf#wrap({
            \ 'source': GetJumps(),
            \ 'sink': function('GoToJump')}))

**Comment by inengch** [comment]
> Here's my solution, inspired by
> https://www.reddit.com/r/vim/comments/a3e5q7
> 
> ```
> function! GetJumps()
>   redir => cout
>   silent jumps
>   redir END
>   return reverse(split(cout, "\n")[1:])
> endfunction
> function! GoToJump(jump)
>     let jumpnumber = split(a:jump, '\s\+')[0]
>     execute "normal " . jumpnumber . "\<c-o>"
> endfunction
> command! Jumps call fzf#run(fzf#wrap({
>         \ 'source': GetJumps(),
>         \ 'sink': function('GoToJump')}))
> ```

How to add the content preview of the jump list result?

**Comment by inengch** [comment]
```
function! GetJumps()
  redir => cout
  silent jumps
  redir END
  return reverse(split(cout, "\n")[1:])
endfunction
function! GoToJump(jump)
    let jumpnumber = split(a:jump, '\s\+')[0]
    execute "normal " . jumpnumber . "\<c-o>"
endfunction
command! Jump call fzf#run(fzf#vim#with_preview(fzf#wrap({
        \ 'source': GetJumps(),
        \ 'sink': function('GoToJump')})))
```
I change the function to this. But it seems that it doesn't work.

**Comment by uestueab** [comment]
The answer provided by @patricknraanes doesn't really jump to the exact position. And having a preview as suggested by @inengch  would be a nice addition. How would one implement these things?

**Comment by uestueab** [comment]
> This works. It requires `bat`. It highlights the line of the jump.
> 
> ```
> function! Jumps()
>   let jumpfile=tempname()
>   let jumpnumfile=tempname()
> 
>   " Get jumps with filename added
>   let jumps = map(reverse(copy(getjumplist()[0])), 
>     \ { key, val -> extend(val, {'name': getbufinfo(val.bufnr)[0].name }) })
> 
>   " Write jumps to temp file with: jumpnumber, line number, line number - half screen, file name
>   let offset = &lines / 2
>   let jumptext = map(copy(jumps), { index, val -> 
>     \ (index + 1)."\t".(val.lnum)."\t".(val.lnum < offset ? 1 : val.lnum - offset)."\t".(val.col+1)."\t".(val.name) })
>   call writefile(jumptext, jumpfile)
> 
>   execute 'silent !fzf --preview "bat {+5} -H {+2} -r {+3}: --color=always" < '.jumpfile.' > '.jumpnumfile
>   call delete(jumpfile)
> 
>   " Read fzf output and goto jump
>   let jumpnumber=readfile(jumpnumfile)
>   call delete(jumpnumfile)
>   if(len(jumpnumber) > 0)
>     let values = split(jumpnumber[0], "\t")
>     execute "e ".values[4]
>     call cursor(str2nr(values[1]), str2nr(values[3]))
>   endif
>   redraw!
> 
> endfunction
> 
> command! Jumps call Jumps()
> ```
> 
> I'm new to vimscript and fzf.vim source. If I have some time later, I'll rewrite it to leverage the fzf.vim API.
> 
> But wait! There's more. Change lists:
> 
> ```
> function! Changes()
>   let changefile=tempname()
>   let changenumfile=tempname()
> 
>   let changes  = reverse(copy(getchangelist()[0]))
> 
>   " Write changes to temp file with: changenumber, line number, line number - 15, file name
>   let offset = &lines / 2
>   let changetext = map(copy(changes), { index, val -> 
>     \ (val.lnum)."\t".(val.lnum < offset ? 1 : val.lnum - offset)."\t".(val.col+1)."\t".trim(getline(val.lnum)) })
>   call writefile(changetext, changefile)
> 
>   execute 'silent !fzf --preview "bat % -H {+1} -r {+2}: --color=always" < '.changefile.' > '.changenumfile
>   call delete(changefile)
> 
>   " Read fzf output and goto change
>   let changenumber=readfile(changenumfile)
>   call delete(changenumfile)
>   if(len(changenumber) > 0)
>     let values = split(changenumber[0], "\t")
>     call cursor(str2nr(values[0]), str2nr(values[2]))
>   endif
>   redraw!
> 
> endfunction
> 
> command! Changes call Changes()
> ```
> 
> Example mappings
> 
> ```
> nnoremap <leader>j :Jumps<cr>zz
> nnoremap <leader>c :Changes<cr>zz
> 
> " add relative moves to jumplist
> nnoremap <expr> j (v:count > 5 ? "m'" . v:count . "j" : "j")
> nnoremap <expr> k (v:count > 5 ? "m'" . v:count . "k" : "k")
> ```

Alright this looks fine. It takes  the whole screen though and also having a some part of the line be displayed on the left pane would be nice

**Comment by mikeslattery** [comment]
> @mikeslattery
> 
> Alright this looks fine. It takes the whole screen though and also having a little part of the line be displayed on the left pane would be nice

@uestueab  I consider the current code a PoC / workaround.   As I said, I'll rewrite this using fzf.vim functions when I get the time.  That will take care of all display-related issues.  I'll base it off `fzf#vim#marks`

**Comment by mikeslattery** [comment]
@uestueab  Improvements!   Not perfect, but closer.

```
function GoTo(jumpline)
  let values = split(a:jumpline, ":")
  echo "e ".values[0]
  call cursor(str2nr(values[1]), str2nr(values[2]))
endfunction

function GetLine(bufnr, lnum)
  let lines = getbufline(a:bufnr, a:lnum)
  if len(lines)>0
    return trim(lines[0])
  else
    return ''
  endif
endfunction

function! Jumps()
  " Get jumps with filename added
  let jumps = map(reverse(copy(getjumplist()[0])), 
    \ { key, val -> extend(val, {'name': getbufinfo(val.bufnr)[0].name }) })
 
  " Write jumps to temp file with: jumpnumber, line number, line number - half screen, file name
  let jumptext = map(copy(jumps), { index, val -> 
      \ (val.name).':'.(val.lnum).':'.(val.col+1).': '.GetLine(val.bufnr, val.lnum) })

  call fzf#run(fzf#vim#with_preview(fzf#wrap({
        \ 'source': jumptext,
        \ 'column': 1,
        \ 'options': ['--delimiter', ':', '--bind', 'alt-a:select-all,alt-d:deselect-all', '--preview-window', '+{2}-/2'],
        \ 'sink': function('GoTo')})))
endfunction

command! Jumps call Jumps()

function! Changes()
  let changefile=tempname()
  let changenumfile=tempname()

  let changes  = reverse(copy(getchangelist()[0]))

  " Write changes to temp file with: changenumber, line number, line number - half screen, file name
  let offset = &lines / 2 - 3
  let changetext = map(copy(changes), { index, val -> 
      \ expand('%').':'.(val.lnum).':'.(val.col+1).': '.GetLine(bufnr('%'), val.lnum) })

  call fzf#run(fzf#vim#with_preview(fzf#wrap({
        \ 'source': changetext,
        \ 'column': 1,
        \ 'options': ['--delimiter', ':', '--bind', 'alt-a:select-all,alt-d:deselect-all', '--preview-window', '+{2}-/2'],
        \ 'sink': function('GoTo')})))
endfunction

command! Changes call Changes()
```

TODOS:
* Add jump numbers to `Jumps`
* Add `>` marker at current jump or change.
* Don't include filename in `Changes`.
* Format similar to `jumps` and `changes`  (maybe.  It's less dense)
* Create a pull request

Some of these are regressions, as I copied how `Rg` works.

**Comment by uestueab** [comment]
@mikeslattery
Amazing, this is all i ever wanted. Thanks mate! Though regarding your todo's the '>' already appears for me, but that might just be my configuration.

**Comment by mikeslattery** [comment]
FYI, I'm not planning on implementing my TODOs or creating a PR.   I've switched to a similar plugin that implements fuzzy find on jumplist.   Hopefully, someone will pick up where I left off.  I solved the hard part, I think.

**Comment by nagakiran** [comment]
@mikeslattery 
Can you share the similar plugin you are referring for fuzzy find on jumplist as that may be helpful for others also?

**Comment by lee-shun** [comment]
Thank you guys' great work! 
I just add some little enhancements for the one who still wants to use fzf.vim...

```vim
" from https://github.com/phongnh/fzf-settings.vim/blob/master/plugin/fzf_settings.vim
" Extra commands

function! s:warn(message) abort
    echohl WarningMsg
    echomsg a:message
    echohl None
    return 0
endfunction

function! s:fzf_bufopen(e) abort
    let list = split(a:e)
    if len(list) < 4
        return
    endif

    let [linenr, col, file_text] = [list[1], list[2]+1, join(list[3:])]
    let lines = getbufline(file_text, linenr)
    let path = file_text
    if empty(lines)
        if stridx(join(split(getline(linenr))), file_text) == 0
            let lines = [file_text]
            let path = bufname('%')
        elseif filereadable(path)
            let lines = ['buffer unloaded']
        else
            " Skip.
            return
        endif
    endif

    execute 'e '  . path
    call cursor(linenr, col)
endfunction

function! s:fzf_yank_sink(e) abort
    let @" = a:e
    echohl ModeMsg
    echo 'Yanked!'
    echohl None
endfunction

function! s:fzf_messages_source() abort
    return split(call('execute', ['messages']), '\n')
endfunction

function! s:fzf_messages(bang) abort
    let s:source = 'messages'
    call fzf#run(fzf#wrap('messages', {
                \ 'source':  s:fzf_messages_source(),
                \ 'sink':    function('s:fzf_yank_sink'),
                \ 'options': '+m --prompt "Messages> "',
                \ }, a:bang))
endfunction

command! -bang Messages call <SID>fzf_messages(<bang>0)

function! s:fzf_open_quickfix_item(e) abort
    let line = a:e
    let filename = fnameescape(split(line, ':\d\+:')[0])
    let linenr = matchstr(line, ':\d\+:')[1:-2]
    let colum = matchstr(line, '\(:\d\+\)\@<=:\d\+:')[1:-2]
    execute 'e ' . filename
    call cursor(linenr, colum)
endfunction

function! s:fzf_quickfix_to_grep(v) abort
    return bufname(a:v.bufnr) . ':' . a:v.lnum . ':' . a:v.col . ':' . a:v.text
endfunction

function! s:fzf_get_quickfix() abort
    return map(getqflist(), 's:fzf_quickfix_to_grep(v:val)')
endfunction

function! s:fzf_quickfix(bang) abort
    let s:source = 'quickfix'
    let items = s:fzf_get_quickfix()
    if len(items) == 0
        call s:warn('No quickfix items!')
        return
    endif
    call fzf#run(fzf#wrap('quickfix', {
                \ 'source': items,
                \ 'sink':   function('s:fzf_open_quickfix_item'),
                \ 'options': '--layout=reverse-list --prompt "Quickfix> "'
                \ }, a:bang))
endfunction

function! s:fzf_get_location_list() abort
    return map(getloclist(0), 's:fzf_quickfix_to_grep(v:val)')
endfunction

function! s:fzf_location_list(bang) abort
    let s:source = 'location_list'
    let items = s:fzf_get_location_list()
    if len(items) == 0
        call s:warn('No location list items!')
        return
    endif
    call fzf#run(fzf#wrap('location_list', {
                \ 'source': items,
                \ 'sink':   function('s:fzf_open_quickfix_item'),
                \ 'options': '--layout=reverse-list --prompt "LocationList> "'
                \ }, a:bang))
endfunction

command! -bang Quickfix call s:fzf_quickfix(<bang>0)
command! -bang LocationList call s:fzf_location_list(<bang>0)

function! s:fzf_yank_register(line) abort
    call setreg('"', getreg(a:line[7]))
    echohl ModeMsg
    echo 'Yanked!'
    echohl None
endfunction

function! s:fzf_get_registers() abort
    return split(call('execute', ['registers']), '\n')[1:]
endfunction

function! s:fzf_registers(bang) abort
    let s:source = 'registers'
    let items = s:fzf_get_registers()
    if len(items) == 0
        call s:warn('No register items!')
        return
    endif
    call fzf#run(fzf#wrap('registers', {
                \ 'source':  items,
                \ 'sink':    function('s:fzf_yank_register'),
                \ 'options': '--layout=reverse-list +m --prompt "Registers> "',
                \ }, a:bang))
endfunction

command! -bang Registers call s:fzf_registers(<bang>0)

" ===
" === Jumps list and Change list
" ===
" from https://github.com/junegunn/fzf.vim/issues/865

function GoTo(jumpline)
  let values = split(a:jumpline, ":")
  execute "e ".values[0]
  call cursor(str2nr(values[1]), str2nr(values[2]))
  execute "normal zvzz"
endfunction

function GetLine(bufnr, lnum)
  let lines = getbufline(a:bufnr, a:lnum)
  if len(lines)>0
    return trim(lines[0])
  else
    return ''
  endif
endfunction

function Getjumps()
    let jumps = []
    let raw_jumps = reverse(copy(getjumplist()[0]))
    for it in raw_jumps
        if bufexists(it.bufnr)
            call add(jumps, it)
        endif
    endfor
    return jumps
endfunction

function! Jumps()
  " Get jumps with filename added
  let tmp_jump = Getjumps()
  if(tmp_jump == [])
        call s:warn('Empty jump list!')
        return
  endif
  let jumps = map(Getjumps(), 
    \ { key, val -> extend(val, {'fname': getbufinfo(val.bufnr)[0].name }) })
 
  let jumptext = map(copy(jumps), { index, val -> 
      \ (val.fname).':'.(val.lnum).':'.(val.col+1).': '.GetLine(val.bufnr, val.lnum) })

  call fzf#run(fzf#vim#with_preview(fzf#wrap({
        \ 'source': jumptext,
        \ 'column': 1,
        \ 'options': ['--delimiter', ':', '--bind', 'alt-a:select-all,alt-d:deselect-all', '--preview-window', '+{2}-/2'],
        \ 'sink': function('GoTo')})))
endfunction

command! Jumps call Jumps()

function! Changes()
  let changes  = reverse(copy(getchangelist()[0]))
  if(changes == [])
        call s:warn('Empty change list!')
        return
  endif

  let changetext = map(copy(changes), { index, val -> 
      \ expand('%').':'.(val.lnum).':'.(val.col+1).': '.GetLine(bufnr('%'), val.lnum) })

  call fzf#run(fzf#vim#with_preview(fzf#wrap({
        \ 'source': changetext,
        \ 'column': 1,
        \ 'options': ['--delimiter', ':', '--bind', 'alt-a:select-all,alt-d:deselect-all', '--preview-window', '+{2}-/2'],
        \ 'sink': function('GoTo')})))
endfunction

command! Changes call Changes()
```

**Comment by bryclee** [comment]
Another implementation, based on the other solutions in the thread, which doesn't modify the jump list when jumping to items, and starts FZF with the current position in the jump list selected:

```vim
function! Jumplist()
  let jumplist = getjumplist()
  let raw_jumps = reverse(copy(jumplist[0]))
  let s:jump_start = len(raw_jumps) - jumplist[1] - 1
  let s:last_jumplist = map(copy(raw_jumps), { index, val ->
        \ (bufname(val.bufnr)) . ':' . (val.lnum) . ':' . (val.col+1) . ': ' . (trim(get(getbufline(val.bufnr, val.lnum), 0, ''))) })

  call fzf#run(fzf#vim#with_preview(fzf#wrap({
        \ 'source': jumps,
        \ 'sink': function('GoToJump'),
        \ 'options': ['--delimiter', ':', '--no-sort', '--bind', 'load:pos(' . (s:jump_start + 1) . ')', '--preview-window', '+{2}-/2'] })))
endfunction
function! GoToJump(jump)
    let pos = index(s:last_jumplist, a:jump)
    let jumpnumber = abs(pos - s:jump_start)
    execute "normal " . jumpnumber . (pos >= s:jump_start ? "\<c-o>" : "\<c-i>")
endfunction
```

**Comment by junegunn** [comment]
Closed by #710
